### PR TITLE
Updater: resource compression

### PR DESCRIPTION
**Pull request by hedger**
# What's new

- toolbox: compress: added callback-based compression API for decompressing large streams at once
- toolbox: compress: added compression API for chunked reads of decompressed data
- toolbox: compress: refactored existing compressor to use callback-based API internally
- toolbox: tar: support for heatshrink compression
- cli: added `storage extract` command for extracting plain .tar and .ths archives at arbitrary paths
- updater: now using .tar.hs for resource bundle
- updater: split resource deployment stage progress into separate cleanup and unpack steps
- unit_tests: tests for all the new stuff
- scripts: added standalone script for packing/unpacking data with Flipper Heathshrink stream headers
- rpc: now uses heatshrink decompression for archives ending with `.ths`

# Verification 

- Run update, `storage extract` in CLI
- idk

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Comment by CookiePLMonster** [comment]
Will the use case here be that the users are expected to compress files off-site and e.g. include compressed assets inside the FAP, or is this in preparation for FAPs compressing assets internally and decompressing during the first-time extraction step?

**Comment by hedger** [comment]
> Will the use case here be that the users are expected to compress files off-site and e.g. include compressed assets inside the FAP, or is this in preparation for FAPs compressing assets internally and decompressing during the first-time extraction step?

Primary objective for this PR is to reduce resource bundle size in update package, improving update speeds.

However, with improved compression APIs, it may be reused in other subsystems.

**Comment by CookiePLMonster** [comment]
> However, with improved compression APIs, it may be reused in other subsystems.

Yeah - considering that FAP asset unpacking is IO bound (as far as I can tell), and those assets can never be referenced directly without unpacking, feels like storing them compressed is a win - less IO reads, and faster FAP installation times.

File: lib/toolbox/compress.c [hunk 1]
```diff
@@ -13,9 +13,17 @@
...
-/** Buffer size for input data */
 #define COMPRESS_ICON_ENCODED_BUFF_SIZE (256u)
 
+static const CompressConfigHeatshrink COMPRESS_CONFIG_HEATSHRINK = {
+    .window_sz2 = COMPRESS_EXP_BUFF_SIZE_LOG,
+    .lookahead_sz2 = COMPRESS_LOOKAHEAD_BUFF_SIZE_LOG,
+    .input_buffer_sz = COMPRESS_ICON_ENCODED_BUFF_SIZE,
+};
+
+const CompressConfigHeatshrink* compress_config_heatshrink_default = &COMPRESS_CONFIG_HEATSHRINK;
```
Comment (skotopes) [inline lib/toolbox/compress.c]: WTF?

File: lib/toolbox/compress.h [hunk 2]
```diff
@@ -100,6 +120,100 @@ bool compress_decode(
...
 
+/** I/O callback for streamed compression/decompression
+ * 
+ * @param context user context
+ * @param buffer buffer to read/write
+ * @param size size of buffer
+ * 
+ * @return number of bytes read/written, 0 on end of stream, negative on error
+ */
+typedef int32_t (*compress_io_cb_t)(void* context, uint8_t* buffer, int32_t size);
```
Comment (skotopes) [inline lib/toolbox/compress.h]: Naming rules, would you kindly?

**Review by skotopes — APPROVED** [review]

**Pull request merged**